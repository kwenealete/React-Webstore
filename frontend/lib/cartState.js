import { createContext, useContext, useState } from 'react';

const LocalStateContext = createContext();
const LocalStateProvider = LocalStateContext.Provider;

function CartStateProvider({ children }) {
    // Our customised custom provider! data (state) will be stored here with functionality (updaters) and made accessible via the consumer

    // Closed cart by default
    const [cartOpen, setCartOpen] = useState(false);

    function toggleCart() {
        setCartOpen(!cartOpen);
    }

    function closeCart() {
        setCartOpen(false);
    }

    function openCart() {
        setCartOpen(true);
    }

    return (
        <LocalStateProvider
            value={{
                cartOpen,
                setCartOpen,
                toggleCart,
                closeCart,
                openCart
            }} >
            {children}
        </LocalStateProvider>
    );
}

// Make a custom hook for accessing the cart local state
function useCart() {
    // Using a consumer to access the local state
    const all = useContext(LocalStateContext);
    return all;
}
export { CartStateProvider, useCart };