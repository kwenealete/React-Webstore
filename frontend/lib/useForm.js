import { useState, useEffect } from 'react';

export default function useForm(initial = {}) {
    //creating a state object for our inputs

    const [inputs, setInputs] = useState(initial);
    const initialValues = Object.values(initial).join('');

    useEffect(() => {
        //it runs when changes occurs on things we are watching
        setInputs(initial);
    }, [initialValues]);

    function handleChange(e) {
        let { value, name, type } = e.target;

        if(type === 'number') {
            value = parseInt(value);
        }

        if(type === 'file') {
            [value] = e.target.files;
        }

        setInputs({
            //copy existing state
            ...inputs,
            [name]: value,
        });
    }

    function resetForm() {
        setInputs(initial);
    }

    function clearForm() {
        const blankState = Object.fromEntries(Object.entries(inputs).map(([key, value]) => [key, '']));
        setInputs(blankState);
    }

    //return things we want to surface from this hook

    return {
        inputs,
        handleChange,
        resetForm,
        clearForm,
    };
}