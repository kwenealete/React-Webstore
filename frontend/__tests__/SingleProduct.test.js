import { render, screen } from '@testing-library/react';
import { MockedProvider } from '@apollo/client/testing';
import SingleProduct, { SINGLE_ITEM_QUERY } from '../components/SingleProduct';
import { fakeItem } from '../lib/testUtils';


const product = fakeItem();

const mocks = [
    {
        //When this query and variable combo is requested
        request: {
            query: SINGLE_ITEM_QUERY,
            variables: { id: '123' },
        },
        //Return this data
        result: {
            data: {
                Product: product,
            },
        },
    },
];

describe('<Single Product/>', () => {
    it('renders with proper data', async() => {
        //Creating some fake data

        const { container, debug } = render(
            <MockedProvider mocks={mocks}>
                <SingleProduct id='123' />
            </MockedProvider>
        );
        await screen.findByTestId('singleProduct');
        expect(container).toMatchSnapshot();
    });

    it('Errors out when an item is not found', async() => {
        const errorMock = [{
            request: {
                query: SINGLE_ITEM_QUERY,
                variables: { id: '123' },
            },

            results: {
                errors: [{ message: 'item not found' }]
            },
        }]; 

        const { container, debug } = render (
            <MockedProvider mocks={errorMock}>
                <SingleProduct id="123" />
            </MockedProvider>
        );
        await screen.findByTestId('graphql-error');
        expect(container).toMatchSnapshot();
        expect(container).toHaveTextContent('Shoot!');
        
    })
});