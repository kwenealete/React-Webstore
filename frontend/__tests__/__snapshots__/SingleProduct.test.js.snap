// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Single Product/> Errors out when an item is not found 1`] = `
<div>
  <div
    class="ErrorMessage__ErrorStyles-sc-11u5fgj-0 cEfbva"
  >
    <p
      data-testid="graphql-error"
    >
      <strong>
        Shoot!
      </strong>
      Mocked response should contain either result or error: {"query":"query SINGLE_ITEM_QUERY($id: ID!) {\\n  Product(where: {id: $id}) {\\n    name\\n    price\\n    description\\n    id\\n    photo {\\n      altText\\n      image {\\n        publicUrlTransformed\\n        __typename\\n      }\\n      __typename\\n    }\\n    __typename\\n  }\\n}"}
    </p>
  </div>
</div>
`;

exports[`<Single Product/> renders with proper data 1`] = `
<div>
  <div
    class="SingleProduct__ProductStyles-sc-1kapz8w-0 esggPb"
    data-testid="singleProduct"
  >
    <img
      alt="dogs are best"
      src="dog.jpg"
    />
    <div
      class="details"
    >
      <h2>
        dogs are best
      </h2>
      <p>
        dogs
      </p>
    </div>
  </div>
</div>
`;
